src/seat/input_method.rs

```diff
@@ -0,0 +1,391 @@
+                imdata.pending_state = InputMethodEventState {
+                    active: imdata.pending_state.active.with_active(),
+                    ..Default::default()
+                };
+            }
+            Event::Deactivate => {
+                imdata.pending_state = Default::default();
+            }
+            Event::SurroundingText { text, cursor, anchor } => {
+                imdata.pending_state = InputMethodEventState {
+                    active: imdata.pending_state.active.with_surrounding_text(),
+                    surrounding: SurroundingText { text, cursor, anchor },
+                    ..imdata.pending_state.clone()
+                }
+            }
+            Event::TextChangeCause { cause } => {
+                imdata.pending_state = InputMethodEventState {
+                    text_change_cause: match cause {
+                        WEnum::Value(cause) => cause,
+                        WEnum::Unknown(value) => {
+                            warn!(
+                                "Unknown `text_change_cause`: {}. Assuming not input method.",
+                                value
+                            );
+                            ChangeCause::Other
+                        }
+                    },
+                    ..imdata.pending_state.clone()
+                }
+            }
+            Event::ContentType { hint, purpose } => {
+                imdata.pending_state = InputMethodEventState {
+                    active: imdata.pending_state.active.with_content_type(),
+                    content_hint: match hint {
+                        WEnum::Value(hint) => hint,
+                        WEnum::Unknown(value) => {
+                            warn!(
+                                "Unknown content hints: 0b{:b}, ignoring.",
+                                ContentHint::from_bits_retain(value)
+                                    - ContentHint::from_bits_truncate(value)
+                            );
+                            ContentHint::from_bits_truncate(value)
+                        }
+                    },
+                    content_purpose: match purpose {
+                        WEnum::Value(v) => v,
+                        WEnum::Unknown(value) => {
+                            warn!("Unknown `content_purpose`: {}. Assuming `normal`.", value);
+                            ContentPurpose::Normal
+                        }
+                    },
+                    ..imdata.pending_state.clone()
+                }
+            }
+            Event::Done => {
+                imdata.pending_state = InputMethodEventState {
+                    active: imdata.pending_state.active.with_done(),
+                    ..imdata.pending_state.clone()
+                };
+                imdata.current_state = imdata.pending_state.clone();
+                imdata.serial += 1;
+                data.handle_done(qh, &input_method, &imdata.current_state)
```

Comment:
```suggestion
                data.handle_done(qh, input_method, &imdata.current_state)
```
Pretty sure this is a needless double borrow